// react-pdf-highlighter/build/components/StickyNote.js
import React, { Component } from "react";
import "../style/StickyNote.css"; // Ensure to create this CSS file for custom styling

class StickyNote extends Component {
    render() {
        const {
            position,
            onClick,
            onMouseOver,
            onMouseOut,
            comment,
            isScrolledTo
        } = this.props;
        const { boundingRect } = position;

        // Styling for the sticky note
        const stickyStyle = {
            position: 'absolute',
            top: boundingRect.top, // Position at the top of the bounding box
            left: boundingRect.right, // Position to the right of the bounding box
            backgroundColor: 'yellow', // Typical sticky note color
            padding: '8px',
            borderRadius: '4px',
            boxShadow: '0 2px 4px rgba(0,0,0,0.2)',
            zIndex: 1000 // Ensure it floats above other content
        };

        return (
            <div
                className={`StickyNote ${isScrolledTo ? "StickyNote--scrolledTo" : ""}`}
                onMouseOver={onMouseOver}
                onMouseOut={onMouseOut}
                onClick={onClick}
                style={stickyStyle}
            >
                {comment ? <div>{comment.text}</div> : null}
            </div>
        );
    }
}

export default StickyNote;
